File: content/LLD/Reference-Code/13. Abstract Factory Pattern/good_example.py
from abc import ABC, abstractmethod


class Starter(ABC):
    @abstractmethod
    def prepare(self):
        pass


class MainCourse(ABC):
    @abstractmethod
    def prepare(self):
        pass


class Dessert(ABC):
    @abstractmethod
    def prepare(self):
        pass


# ============= NORTH INDIAN DISHES =============


class PaneerTikka(Starter):
    def prepare(self):
        print("Preparing Paneer Tikka (North Indian Starter)")


class ButterChicken(MainCourse):
    def prepare(self):
        print("Preparing Butter Chicken (North Indian Main Course)")


class GulabJamun(Dessert):
    def prepare(self):
        print("Preparing Gulab Jamun (North Indian Dessert)")


# ============= SOUTH INDIAN DISHES =============


class MeduVada(Starter):
    def prepare(self):
        print("Preparing Medu Vada (South Indian Starter)")


class Dosa(MainCourse):
    def prepare(self):
        print("Preparing Dosa (South Indian Main Course)")


class Payasam(Dessert):
    def prepare(self):
        print("Preparing Payasam (South Indian Dessert)")


# ============= CHINESE DISHES =============


class SpringRolls(Starter):
    def prepare(self):
        print("Preparing Spring Rolls (Chinese Starter)")


class FriedRice(MainCourse):
    def prepare(self):
        print("Preparing Fried Rice (Chinese Main Course)")


class FortuneCookie(Dessert):
    def prepare(self):
        print("Preparing Fortune Cookie (Chinese Dessert)")


class CuisineFactory(ABC):
    @abstractmethod
    def create_starter(self) -> Starter:
        pass

    @abstractmethod
    def create_main_course(self) -> MainCourse:
        pass

    @abstractmethod
    def create_dessert(self) -> Dessert:
        pass


class NorthIndianCuisine(CuisineFactory):
    def create_starter(self) -> Starter:
        return PaneerTikka()

    def create_main_course(self) -> MainCourse:
        return ButterChicken()

    def create_dessert(self) -> Dessert:
        return GulabJamun()


class ChineseCuisine(CuisineFactory):
    def create_starter(self) -> Starter:
        return SpringRolls()

    def create_main_course(self) -> MainCourse:
        return FriedRice()

    def create_dessert(self) -> Dessert:
        return FortuneCookie()


class RestaurantService:
    def __init__(self, factory: CuisineFactory):
        self.__factory = factory

    def create_meal(self):
        starter = self.__factory.create_starter()
        main_course = self.__factory.create_main_course()
        dessert = self.__factory.create_dessert()

        starter.prepare()
        main_course.prepare()
        dessert.prepare()

    def change_cuisine(self, new_factory: CuisineFactory):
        self.__factory = new_factory


north_indian_cuisine = NorthIndianCuisine()
restaurant_service = RestaurantService(north_indian_cuisine)
restaurant_service.create_meal()

chinese = ChineseCuisine()
restaurant_service.change_cuisine(chinese)
restaurant_service.create_meal()

File: content/LLD/Reference-Code/13. Abstract Factory Pattern/GoodExample.java
public class GoodExample {
    interface Starter { void prepare(); }
    interface MainCourse { void prepare(); }
    interface Dessert { void prepare(); }

    // North Indian
    static class PaneerTikka implements Starter {
        public void prepare() { System.out.println("Preparing Paneer Tikka (North Indian Starter)"); }
    }
    static class ButterChicken implements MainCourse {
        public void prepare() { System.out.println("Preparing Butter Chicken (North Indian Main Course)"); }
    }
    static class GulabJamun implements Dessert {
        public void prepare() { System.out.println("Preparing Gulab Jamun (North Indian Dessert)"); }
    }

    // South Indian
    static class MeduVada implements Starter {
        public void prepare() { System.out.println("Preparing Medu Vada (South Indian Starter)"); }
    }
    static class Dosa implements MainCourse {
        public void prepare() { System.out.println("Preparing Dosa (South Indian Main Course)"); }
    }
    static class Payasam implements Dessert {
        public void prepare() { System.out.println("Preparing Payasam (South Indian Dessert)"); }
    }

    // Chinese
    static class SpringRolls implements Starter {
        public void prepare() { System.out.println("Preparing Spring Rolls (Chinese Starter)"); }
    }
    static class FriedRice implements MainCourse {
        public void prepare() { System.out.println("Preparing Fried Rice (Chinese Main Course)"); }
    }
    static class FortuneCookie implements Dessert {
        public void prepare() { System.out.println("Preparing Fortune Cookie (Chinese Dessert)"); }
    }

    interface CuisineFactory {
        Starter createStarter();
        MainCourse createMainCourse();
        Dessert createDessert();
    }

    static class NorthIndianCuisine implements CuisineFactory {
        public Starter createStarter() { return new PaneerTikka(); }
        public MainCourse createMainCourse() { return new ButterChicken(); }
        public Dessert createDessert() { return new GulabJamun(); }
    }

    static class ChineseCuisine implements CuisineFactory {
        public Starter createStarter() { return new SpringRolls(); }
        public MainCourse createMainCourse() { return new FriedRice(); }
        public Dessert createDessert() { return new FortuneCookie(); }
    }

    static class RestaurantService {
        private CuisineFactory factory;

        public RestaurantService(CuisineFactory factory) {
            this.factory = factory;
        }

        public void createMeal() {
            Starter starter = this.factory.createStarter();
            MainCourse mainCourse = this.factory.createMainCourse();
            Dessert dessert = this.factory.createDessert();

            starter.prepare();
            mainCourse.prepare();
            dessert.prepare();
        }

        public void changeCuisine(CuisineFactory newFactory) {
            this.factory = newFactory;
        }
    }

    public static void main(String[] args) {
        NorthIndianCuisine northIndianCuisine = new NorthIndianCuisine();
        RestaurantService restaurantService = new RestaurantService(northIndianCuisine);
        restaurantService.createMeal();

        ChineseCuisine chinese = new ChineseCuisine();
        restaurantService.changeCuisine(chinese);
        restaurantService.createMeal();
    }
}
